File: erlang/p021.erl
-module(p021).

-export([start/0, f/2]).

d(N) ->
  lists:sum(utils:divisors(N)).

f(I, J) ->
  lists:sum([ X || X <- lists:seq(I, J), X /= d(X), X == d(d(X))]).

launch(Processes) ->
  lists:sum([ rpc:yield(P) || P <- launch(1, 10000 div Processes, 10000) ]).

launch(I, _, Max) when I > Max ->
  [];

launch(I, Increment, Max) ->
  launch(I + Increment, Increment, Max) ++ [rpc:async_call(node(), ?MODULE, f, [I, I + Increment - 1])].

start() ->
  io:format("~p ~n", [launch(8 * erlang:system_info(logical_processors_available))]).

File: go/src/p021/p021.go
package main

import (
    "fmt"
    "utils"
)

func main() {
    sum := 0
    for a := 2; a < 10000; a++ {
        b := d(a)
        if a != b && a == d(b) {
            sum += a
        }
    }
    fmt.Println(sum)
}

func d(n int) int {
    sum := -n
    for _, i := range(utils.Divisors(uint64(n))) {
        sum += int(i)
    }
    return sum
}
